<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Binmarkers by lileiting</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Binmarkers</h1>
      <h2 class="project-tagline">Bin SNP markers based on their physical positions</h2>
      <a href="https://github.com/lileiting/BinMarkers" class="btn">View on GitHub</a>
      <a href="https://github.com/lileiting/BinMarkers/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/lileiting/BinMarkers/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="binmarkers" class="anchor" href="#binmarkers" aria-hidden="true"><span class="octicon octicon-link"></span></a>BinMarkers</h1>

<h2>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span class="octicon octicon-link"></span></a>Overview</h2>

<p>BinMarkers is a program for creating bin markers from a large set of SNP markers based on their positions on scaffolds/contigs. The purpose of this program is to reduce the number of markers for constructing genetic maps using <a href="http://www.kyazma.nl/index.php/mc.JoinMap/">Joinmap</a>, which limited to about six thousands markers.</p>

<ul>
<li>Author: <a href="https://github.com/lileiting">Leiting Li</a>, <a href="https://github.com/freemao">Chenyong Miao</a>
</li>
<li>Email: <a href="mailto:lileiting@gmail.com">lileiting@gmail.com</a>
</li>
<li>LICENCE: <a href="http://opensource.org/licenses/bsd-license.php">BSD</a>
</li>
</ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<pre><code>git clone https://github.com/lileiting/BinMarkers.git
</code></pre>

<h2>
<a id="dependencies" class="anchor" href="#dependencies" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dependencies</h2>

<p><a href="http://www.perl.org">Perl</a> programming language and <a href="http://pdl-stats.sourceforge.net/Distr.htm#pmf_binomial">pmf_binomial</a> function in Perl module <a href="https://metacpan.org/pod/PDL::Stats">PDL::Stats</a></p>

<p>Dependencies of PDL::Stats</p>

<ul>
<li>
<a href="https://metacpan.org/pod/PDL">PDL</a>, The Perl Data Language</li>
<li>
<a href="http://www.gnu.org/software/gsl/">GSL</a>, GNU Scientific Library</li>
</ul>

<p>How to install PDL::Stats</p>

<pre><code>perl -MCPAN -e shell
install PDL
install PDL::Stats
</code></pre>

<p>or use cpan</p>

<pre><code>cpan PDL
cpan PDL::Stats
</code></pre>

<p>or (if without permission to install)</p>

<pre><code>sudo cpan PDL
sudo cpan PDL::Stats
</code></pre>

<p>or use cpanp</p>

<pre><code>cpanp -i PDL
cpanp -i PDL::Stats
</code></pre>

<p>or use cpanm</p>

<pre><code>cpanm PDL
cpanm PDL::Stats
</code></pre>

<p>How to install GSL</p>

<p>Use <a href="http://brew.sh">brew</a> in Mac OS</p>

<pre><code>brew install gsl
</code></pre>

<p>Use yum in CentOS</p>

<pre><code>sudo yum install gsl
sudo yum install gsl-devel
</code></pre>

<p>Use apt-get in Ubuntu</p>

<pre><code>sudo apt-get install libgsl0ldbl
sudo apt-get install gsl-bin libgsl0-dev
</code></pre>

<h2>
<a id="how-to-use-binmarkers" class="anchor" href="#how-to-use-binmarkers" aria-hidden="true"><span class="octicon octicon-link"></span></a>How to use BinMarkers</h2>

<p>Input data format</p>

<p>Input data should be a matrix, one marker per line, one individual per column.
The first line is the title containing individual names, the first column is 
marker names.</p>

<p>Marker names should contain two parts. First is the scaffold/contig/chromosome 
name, second is the position, and they were joined by dash or underscore, e.g.
scaffold1_12345, scaffold1-12345</p>

<p>The default genotypes are "a" for 0/0, "h" for 0/1, "b" for 1/1, "-" for 
missingthese could be customized using -0, -1, -2 and -m options.</p>

<p>Examples</p>

<pre><code>perl binmarkers.pl markers.map -t 5
perl binmarkers.pl markers.map --threshold 5
perl binmarkers.pl markers.map -t 5 -0 a -1 h -2 b -m '-'
perl binmarkers.pl markers.map --threshold=5 --letter_for_0_0=a --letter_for_0_1=h --letter_for_1_1=b --missing='-'
perl binmarkers.pl markers.map -t 5 -0 a -1 h -2 b -m '-' --error_rate_for_0_0 0.04 --error_rate_for_0_1 0.03 --error_rate_for_1_1 0.01
</code></pre>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/lileiting/BinMarkers">Binmarkers</a> is maintained by <a href="https://github.com/lileiting">lileiting</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
